MIPs/mip-3.md

```diff
@@ -0,0 +1,184 @@
+
+- `send_to`: (Optional) Allows integration of the implementation of the swap send to feature.
+
+MetaMask will interpret the method call and perform the necessary validations and operations to initiate the token swap.
+
+## Caveats
+
+The implementation of `wallet_swapAsset` will be new and this could have some time for adoption.
+
+## Implementation
+
+The MetaMask team will be responsible for implementing the JSON RPC `wallet_swapAsset` method. This will involve updating the wallet's internal logic.
+
+## Developer Adoption Considerations
+
+Considering that this is a new json rpc method, developers should take the following aspects into account when adopting this proposal:
+
+1. Backward compatibility: The introduction of the `wallet_swapAsset` method does not affect existing methods or functionality. Developers can integrate this method into their dApps without needing to modify existing code.
+
+2. Ease of adoption: The `wallet_swapAsset` method is designed to be straightforward to use. Developers simply need to provide the necessary parameters (source token details, destination token details) in the method call. The wallet handles the rest of the process, including validation, token swap operation, and error handling.
+
+3. User Experience: By using the `wallet_swapAsset` method, developers can improve the user experience of their dApps. Users can perform token swaps without leaving the dApp, making the process more seamless and convenient.
+
+4. Error handling: Dapp Developers should implement appropriate error handling when using the `wallet_swapAsset` method. The method can return various errors, such as invalid parameters or unsupported operations, and these should be handled gracefully in the dApp.
+   On MetaMask side we intent to validate every property of the request and return an approriate answer to the dapp.
+   Currently we throw this errors:
+
+- **Undefined Parameters**: Parameters like `from`, `to`, `user_address`, etc. are essential for the `wallet_swapAsset` method to work correctly. If not provided, the `validateParams` function would throw an error saying `"${property} property of ${name} is not defined"`. (This validation will change on the future when we allow multiple swap tokens, since the architecture of this rpc method allows it.)
+
+  ```markdown
+  validateParams(from[0], [amount, token_address], 'from');
+  validateParams(to, [token_address], 'to');
+  ```
+
+- **Invalid User Address**: If a non-existent address is provided in `user_address`, then the method will throw an error 'This address does not exist'.
+
+  ```markdown
+  if (!dappConnectedAccount) {
+  throw ethErrors.rpc.invalidParams('This address does not exist');
+  }
+  ```
+
+- **Unsupported Multiple Tokens Swap**: As of now, it supports only a single token swap. In case of multiple tokens provided in the `from` object, an 'Currently we de not support multiple tokens swap' error is thrown.
```

Comment:
```suggestion
- **Unsupported Multiple Tokens Swap**: As of now, it supports only a single token swap. In case of multiple tokens provided in the `from` object, an 'Currently we do not support multiple tokens swap' error is thrown.
```